### ISSUE TITLE
Fallback to PyTorch implementation for bf16 MultiScaleDeformableAttention

### ISSUE THREAD

**Issue by woldier** [VBTI-development/onedl-mmcv#44]
### Description

I am trying to accelerate Mask2Former training with AMP. However, when using the default AMP setting with `float16`, the training loss may become `NaN`.

To avoid the numerical instability caused by `float16`, I tried to switch AMP precision to `bfloat16`:

```python
optim_wrapper = dict(
    type='AmpOptimWrapper',
    dtype='bfloat16'
)
````

However, this causes an error in `MultiScaleDeformableAttnFunction`, because the current CUDA implementation does not support `bf16`.

The related code is here:
https://github.com/VBTI-development/onedl-mmcv/blob/3021df8f1e30dcf8daf3a861d3f21ec8ff00a6b5/mmcv/ops/multi_scale_deform_attn.py#L367-L376

### Problem

When training Mask2Former with AMP:

1. Using default `float16` AMP may lead to `NaN` loss.
2. Switching AMP dtype to `bfloat16` can improve numerical stability.
3. However, the custom CUDA op used by `MultiScaleDeformableAttnFunction` does not support `bf16`, which causes a runtime error.

### Expected behavior

When the input tensor dtype is `bfloat16`, `MultiScaleDeformableAttention` should avoid calling the unsupported CUDA op and instead fall back to the PyTorch implementation.

For example, the logic could be:

```python
 if ((IS_CUDA_AVAILABLE and value.is_cuda) 
         or (IS_MLU_AVAILABLE and value.is_mlu) 
         or (IS_MUSA_AVAILABLE and value.is_musa) 
         or (IS_NPU_AVAILABLE and value.device.type == 'npu'))
          and value.dtype != torch.bfloat16: # fall back to the PyTorch implementation.
     output = MultiScaleDeformableAttnFunction.apply( 
         value, spatial_shapes, level_start_index, sampling_locations, 
         attention_weights, self.im2col_step) 
 else: 
     output = multi_scale_deformable_attn_pytorch( 
         value, spatial_shapes, sampling_locations, attention_weights) 
```

This would allow training with:

```python
optim_wrapper = dict(
    type='AmpOptimWrapper',
    dtype='bfloat16'
)
```

without hitting the unsupported bf16 CUDA path.

### Motivation

`bfloat16` is useful for AMP training because it has a larger exponent range than `float16`, which can reduce the risk of overflow and `NaN` loss during training.

For Mask2Former and other models that rely on multi-scale deformable attention, supporting a bf16 fallback path would make AMP training more stable and practical, while avoiding changes to the CUDA kernel itself.

### Proposed solution

I would like to submit a PR to add a dtype check in `MultiScaleDeformableAttention`.

The PR would not add bf16 support to the CUDA operator directly. Instead, when the input dtype is `torch.bfloat16`, it would use the existing PyTorch implementation as a fallback.

The PR would aim to:

* Detect when the input dtype is `torch.bfloat16`.
* Use `multi_scale_deformable_attn_pytorch` for bf16 inputs.
* Keep the current CUDA op path unchanged for `float32` and `float16`.
* Preserve backward compatibility.
* Add or update tests to cover the bf16 fallback behavior.

Please let me know whether this fallback strategy is acceptable before I start the implementation.

**Comment by lauriebax** [comment]
Hi, thanks for opening this issue. It sounds like a good strategy, but I have some questions:
- will using 'bfloat16' in the 'AmpOptimWrapper' affect other custom layers? Would they need a similar fix?
- I've been looking into the function a bit and it says it needs the custom implementation because of it otherwise will mix fp32 and fp16 dtypes. It is possible your fix will introduce type errors related to this. That being said, if you make tests and it works, it is probably fine.

**Comment by woldier** [comment]
Thanks for the feedback.

- Yes, using `bfloat16` in `AmpOptimWrapper` could affect other custom layers as well, especially custom CUDA ops that do not support bf16. So far, I only encountered the issue in `MultiScaleDeformableAttnFunction`.

- My intention is not to change the existing CUDA path for `float32` or `float16`, but only to fall back to the PyTorch implementation when the input dtype is `torch.bfloat16`, since the CUDA op currently does not support bf16.

My main motivation is to support AMP training for methods such as Mask2Former, where using the default `float16` AMP precision may lead to `NaN` losses during training. In this case, `bfloat16` provides a practical alternative for AMP acceleration with better numerical stability.

This proposed change not affect other methods that already train correctly with `float16` AMP. 
The bf16 fallback would only be triggered when the input dtype is `torch.bfloat16`; otherwise, the existing `float32` / `float16` CUDA paths remain unchanged.

Therefore, for methods that do not suffer from `NaN` losses under `float16` AMP, their training behavior remain the same.
